File: lib/rbac.ts
import type { Role } from "@/lib/types";

export const ROLES = ["super_admin", "admin", "teknisi"] as const;

export const ROLE_LABELS: Record<Role, string> = {
  super_admin: "Super Admin",
  admin: "Admin",
  teknisi: "Teknisi",
};

export type Permissions = {
  role: Role;
  canViewFinance: boolean;
  canMutatePsb: boolean;
  canMutateFinance: boolean;
  canManageUsers: boolean;
};

export function parseRole(value: unknown): Role {
  if (value === "super_admin") return "super_admin";
  if (value === "admin") return "admin";
  return "teknisi";
}

export function permissionsFor(role: Role): Permissions {
  return {
    role,
    canViewFinance: role === "super_admin" || role === "admin",
    canMutatePsb: role === "super_admin" || role === "teknisi",
    canMutateFinance: role === "super_admin",
    canManageUsers: role === "super_admin",
  };
}

export const ROLE_HINTS: Record<Role, string> = {
  super_admin: "Akses penuh, termasuk keuangan dan pengaturan pengguna",
  admin: "Hanya melihat data, tidak bisa tambah/edit/hapus",
  teknisi: "Input/edit/hapus PSB; data keuangan tersembunyi",
};

File: mobile/lib/rbac.dart
enum AppRole { superAdmin, admin, teknisi }

AppRole parseRole(String? value) {
  switch (value) {
    case 'super_admin':
      return AppRole.superAdmin;
    case 'admin':
      return AppRole.admin;
    default:
      return AppRole.teknisi;
  }
}

String roleLabel(AppRole role) {
  switch (role) {
    case AppRole.superAdmin:
      return 'Super Admin';
    case AppRole.admin:
      return 'Admin';
    case AppRole.teknisi:
      return 'Teknisi';
  }
}

class Permissions {
  final AppRole role;
  final bool canMutatePsb;
  final bool canViewFinance;
  final bool canManageUsers;

  const Permissions({
    required this.role,
    required this.canMutatePsb,
    required this.canViewFinance,
    required this.canManageUsers,
  });

  factory Permissions.forRole(AppRole role) {
    return Permissions(
      role: role,
      canMutatePsb: role == AppRole.superAdmin || role == AppRole.teknisi,
      canViewFinance: role == AppRole.superAdmin || role == AppRole.admin,
      canManageUsers: role == AppRole.superAdmin,
    );
  }
}
